src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -830,6 +830,19 @@ private void ParseProject(string firstLine)
                         line = ReadLine();
                     }
                 }
+                else if (line.StartsWith("Project(", StringComparison.Ordinal))
+                {
+                    // Another Project spotted instead of EndProject for the current one - solution file is malformed
+                    string warning = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(out _, out _, "Shared.InvalidProjectFile",
+                        _solutionFile, proj.ProjectName);
+                    SolutionParserWarnings.Add(warning);
+
+                    // The line with new project is already read and we can't go one line back - we have no choice but to recursively parse spotted project
+                    ParseProject(line);
```

Comment:
A small part of me wants to move this to the end of the function. Shouldn't change the behavior, but I think it would be slightly more coherent—deal with one project, then start with the next, rather than interrupting one project to work on the next. As an additional benefit, if C# ever becomes tail recursive, that would limit the number of stack frames we would use in the worst case. 😄

Reply:
@Forgind Parsing the nested project should happen only if `line.StartsWith("Project(", StringComparison.Ordinal)`, so moving it to the end of the whole `ParseProject()` method would result in recursively reading project no matter if this `else if` condition is met.
Of course we can resolve that by setting a flag in `else if` statement, which would be checked at the end of that function - is this what you suggest, or there's something I don't understand.
Please provide me with more details, or perhaps code suggestion.

Reply:
You could also check if line?.StartsWith("Project(", StringComparison.Ordinal)), but I don't think it's worth it. Probably best to just leave it where it is.